Azure - Added functionality to keyvault resource to include the keys in each keyvault resource returned.

bclements opened the pull request:
This will allow us to filter on the keys as well as any key attributes such as "create".

Also added the keyvault client package to setup.py to enable key api functionality.

logachev commented:
I think we shouldn't augment KeyVault resource with all its keys.

I would suggest to create a new KeyVaultKey resource instead. 
It can help to write more granular filters, display specific keys that violate the policy, not just a list of KeyVaults with bad keys.

aluong commented:
You could add a filter that only when invoked will augment KeyVault with the keys. Then the keys will be available on the KeyVault resource for you to perform actions on.

bclements commented:
I've updated the pull with a new commit that moves the keyvault key functionality to it's own resource. I also added a key resource filter that will filter on age.

logachev commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -114,3 +119,48 @@ def enhance_policies(self, access_policies):
+
+
+@resources.register('key')
+class KeyVaultKey(ArmResourceManager):
```
Let's move it to its own file

kapilt replied:
that seems a little extraneous .. modules as class groupings by functionality is quite normal in python.

logachev commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -114,3 +119,48 @@ def enhance_policies(self, access_policies):
         return access_policies
+
+
+@resources.register('key')
```
Can we name it keyvault-key?

bclements replied:
Done, will be in next commit.

logachev commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -114,3 +119,48 @@ def enhance_policies(self, access_policies):
+        s = Session(resource='https://' + url_base)
+        self.keyvault_client = KeyVaultClient(s.get_credentials())
+        r = []
+        for index, item in enumerate(resources):
```
You can do for item in resources:

and then use item instead of resources[index]

bclements replied:
Done, will be in next commit.

logachev commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -114,3 +119,48 @@ def enhance_policies(self, access_policies):
+        enum_spec = ('vaults', 'list', None)
+
+    def augment(self, resources):
+        url_base = 'vault.azure.net'
```
Please move `https://vault.azure.net` to the constants file.

bclements replied:
Done, will be in next commit.

logachev commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -114,3 +119,48 @@ def enhance_policies(self, access_policies):
+
+@KeyVaultKey.filter_registry.register('age')
+class KeyAgeFilter(AgeFilter):
+    date_attribute = "created"
```
I though scenario is to find non-rotated keys.. In that case I assume you're interested in `updated` field or am I wrong?

kapilt replied:
eek, specific agefilters should be scrapped in general for value filters which support it generically with value_type: age on any attribute.

bclements replied:
Removed, the value filter was working fine but I wasn't sure I should do a specific age filter or not (since I saw legacy age filters elsewhere) so just included it. Removing per your request. It will not be included in the next commit.

logachev commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -13,15 +13,19 @@
 # limitations under the License.
```
Please add tests for the new resource, new filter & update docs.

bclements replied:
Will be included in next commit.

aluong replied:
Please update the docs, this is a nice feature. =)

aluong commented on tools/c7n_azure/setup.py:
```diff
@@ -67,6 +67,7 @@
                       "azure-mgmt-subscription",
                       "azure-mgmt-eventgrid>=2.0.0rc1",
                       "azure-graphrbac",
+                      "azure-keyvault",
```
Why do we need this and azure-mgmt-keyvault?

logachev replied:
I think azure-mgmt-keyvault can only perform CRUD with KV resource.. But we need azure-keyvault for operations with keys.

bclements replied:
@logachev is right. I would prefer to use one client rather than 2 different clients but that's how the sdk is providing access to resources right now.

kapilt commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -114,3 +119,48 @@ def enhance_policies(self, access_policies):
+
+            except Exception as e:
+                self.log.error(e.message + ' on keyvault: ' + resources[index]["id"])
+                pass
```
hmm.. why should this pass instead of raising?

bclements replied:
Because I may not have permissions to a storage account that's in my current subscription and I will get auth access errors...similar to s3 auth errors if I don't have permissions as I loop through them. I think it's perfectly fine and ok if you don't have access to a storage account so logging the error message with a pass is ok. discussed with @thisisshi  and he was ok with how I wrote it. Let me know if this is a show stopper to getting this pull approved once I have my final commits in and I can comment further.

thisisshi replied:
rather than doing the try/except on a generic Exception we should be catching on specific ones here.

kapilt replied:
at minimum this try/except should be just around the api call, and catch a specific api client exception, not generic Exception. if an error is persistent wrt to access, i tend to think the policy should fail, unless its explicit about the type of error its catching (ie permission denied) and ideally that behavior is documented. in general we don't catch permission denied, because it means the policy will never work without human intervention, and the  best way to get that is for the policy to error out (error metrics, etc).

logachev commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -114,3 +119,82 @@ def enhance_policies(self, access_policies):
+                vkeys = self.keyvault_client.get_keys(vault_url)
+                for vkey in vkeys:
+                    keyname = vkey.kid.split("/")[-1:]
+                    keyname = keyname[0]
```
nit: keyname = KeyVaultId.parse_key_id(vkey.kid).name

logachev commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -114,3 +119,82 @@ def enhance_policies(self, access_policies):
+            try:
+                vcerts = self.keyvault_client.get_certificates(vault_url)
+                for vcert in vcerts:
+                    certname = vcert.id.split("/")[-1:]
```
nit: certname = KeyVaultId.parse_certificate_id(vkey.kid).name

logachev commented on tools/c7n_azure/tests/test_keyvault-cert.py:
```diff
@@ -0,0 +1,36 @@
+        super(KeyVaultCertTest, self).setUp()
+
+    @arm_template('keyvault-cert.json')
+    def test_find_by_name(self):
```
If you compare the cert age, you might need to mock datetime to return some fixed date.. Otherwise your condition lt 4 can fail based on the run date.

logachev commented on tools/c7n_azure/tests/test_keyvault-key.py:
```diff
@@ -0,0 +1,36 @@
+        super(KeyVaultKeyTest, self).setUp()
+
+    #@arm_template('keyvault-key.json')
+    def test_find_by_name(self):
```
same as another test

kapilt commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -114,3 +121,89 @@ def enhance_policies(self, access_policies):
+                              'created': vkey.attributes.created,
+                              'updated': vkey.attributes.updated,
+                              'tags': vkey.tags})
+            except KeyVaultErrorException as e:
```
there isn't any network call that would generate this exception in the try/except block. parse_key_id is a client local string parse that raises valueerror as well.

https://github.com/Azure/azure-sdk-for-python/blob/master/azure-keyvault/azure/keyvault/key_vault_id.py#L270

kapilt commented on tools/c7n_azure/c7n_azure/resources/key_vault.py:
```diff
@@ -114,3 +121,89 @@ def enhance_policies(self, access_policies):
+                              'updated': vcert.attributes.updated,
+                              'tags': vcert.tags})
+
+            except KeyVaultErrorException as e:
```
same comment applies to this try/except block

aluong commented:
Duplicate - will be addressed with #3897
